cmake/modules/AddSwift.cmake

```diff
@@ -902,3 +902,17 @@ function(set_swift_llvm_is_available name)
   target_compile_definitions(${name} PRIVATE
     $<$<COMPILE_LANGUAGE:C,CXX,OBJC,OBJCXX>:SWIFT_LLVM_SUPPORT_IS_AVAILABLE>)
 endfunction()
+
+# Given the conditions to check to apply the linker flags
+# are not trivial and can be easily forgotten,
+# we preferred to provide a function to configure the targets
+# that on Darwin can skip the IR optimizations when
+# linking with ThinLTO
+function(target_thinlto_ld64_add_flto_codegen_only target)
+  if (SWIFT_TOOLS_ENABLE_LTO STREQUAL "thin" AND
+      LLVM_LINKER_IS_LD64 AND
+      SWIFT_TOOLS_LD64_LTO_CODEGEN_ONLY_FOR_SUPPORTING_TARGETS
+      )
+    target_link_options(${target} PRIVATE "LINKER:-flto-codegen-only")
```

Comment:
I think that a function for this is pretty overkill.  You could just inline this as:
```
target_link_options(... PRIVATE
  $<$<AND:$<BOOL:${LLVM_LINKER_IS_LD64}>,$<BOOL:${SWIFT_TOOLS_LD64_LTO_CODEGEN_ONLY_FOR_SUPPORTING_TARGETS}>,$<STREQUAL:${SWIFT_TOOLS_ENABLE_LTO},thin>>:LINKER:-flto-codegen-only>")
```

Furthermore, these tools would always go through `swift_host_tools`, so we should really just move this invocation to that site.

Reply:
@compnerd can you elaborate on the 2nd thing you are saying. Also I think the reason that this isn't inlined is that it provides a name for people who aren't familiar with generator expressions.

Reply:
I've tried to incorporate the feedback for `swift-ide-test` in 5717f75cbfb2e08925ae5881eb113f7c0d0d3726 so that's easier to compare and discuss the two approaches.

Here are my notes:

* when configuring a host tool I don't see a major different between the two approaches -- even if we go through `add_swift_host_tool` we still need to add a flag to request it
* for me long generator expressions are hard to read, especially when they start nesting contructs -- in fact in this case I resorted breaking the one I wrote on multiple lines, which in turns tempts me to move that logic to a separate function, to reduce the length of `add_swift_host_tool` and to signal this code is taking care of a specific concern at a different level of abstraction
* when debugging using `--trace-expand`, generator expressions are a little bit harder for me to follow since they are not evaluated during configuration but during generation. In particular, with generator expressions the options are seemingly "always added" and an additional check in the build system files is needed, while in the original approach it's clearer to see that we are not taking the branch that adds the options

```
# generator expression approach
swift/cmake/modules/AddSwift.cmake(924):  if(ASHT_THINLTO_LD64_ADD_FLTO_CODEGEN_ONLY )
swift/cmake/modules/AddSwift.cmake(925):  string(CONCAT lto_codegen_only_link_options $<$<AND: $<BOOL:YES>, $<BOOL:TRUE>, $<STREQUAL:thin,thin> >: LINKER:-flto-codegen-only > )
swift/cmake/modules/AddSwift.cmake(933):  target_link_options(swift-ide-test PRIVATE $<$<AND:$<BOOL:YES>,$<BOOL:TRUE>,$<STREQUAL:thin,thin>>:LINKER:-flto-codegen-only> )
<snip>
# original approach
swift/tools/swift-remoteast-test/CMakeLists.txt(16):  target_thinlto_ld64_add_flto_codegen_only(swift-remoteast-test )
swift/cmake/modules/AddSwift.cmake(978):  if(SWIFT_TOOLS_ENABLE_LTO STREQUAL thin AND LLVM_LINKER_IS_LD64 AND SWIFT_TOOLS_LD64_LTO_CODEGEN_ONLY_FOR_SUPPORTING_TARGETS )
swift/cmake/modules/AddSwift.cmake(982):  target_link_options(swift-remoteast-test PRIVATE LINKER:-flto-codegen-only )
```

(Please ignore that in the original approach the variables in the conditions are not expanded, this is a result of the particular syntax I used)

Reply:
Quick update: I discussed this off line with @gottesmm and I will follow Saleem suggestions -- I will take this as a chance to practice with generator expressions.